File: client-api/src/main/java/org/ostelco/topup/api/resources/SubscriptionResource.java
package org.ostelco.topup.api.resources;

import io.dropwizard.auth.Auth;
import io.vavr.control.Either;
import lombok.AllArgsConstructor;
import lombok.NonNull;
import org.ostelco.prime.client.api.model.SubscriptionStatus;
import org.ostelco.topup.api.auth.AccessTokenPrincipal;
import org.ostelco.topup.api.core.Error;
import org.ostelco.topup.api.db.SubscriberDAO;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.Response;

/**
 * Subscriptions API.
 *
 */
@AllArgsConstructor
@Path("/subscription")
public class SubscriptionResource extends ResourceHelpers {

    @NonNull
    private final SubscriberDAO dao;

    @GET
    @Path("status")
    @Produces({"application/json"})
    public Response getSubscription(@Auth AccessTokenPrincipal token) {
        if (token == null) {
            return Response.status(Response.Status.UNAUTHORIZED)
                .build();
        }

        Either<Error, SubscriptionStatus> result = dao.getSubscriptionStatus(token.getName());

        return result.isRight()
            ? Response.status(Response.Status.OK)
                 .entity(getSubscriptionStatusAsJson(result.right().get()))
                 .build()
            : Response.status(Response.Status.NOT_FOUND)
                 .build();
    }
}

File: client-api/src/main/kotlin/org/ostelco/prime/client/api/resources/SubscriptionResource.kt
package org.ostelco.prime.client.api.resources

import io.dropwizard.auth.Auth
import org.ostelco.prime.auth.AccessTokenPrincipal
import org.ostelco.prime.client.api.store.SubscriberDAO
import org.ostelco.prime.jsonmapper.asJson
import org.ostelco.prime.model.Identity
import javax.ws.rs.GET
import javax.ws.rs.Path
import javax.ws.rs.Produces
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response
import kotlin.DeprecationLevel.WARNING

/**
 * Subscriptions API.
 *
 */

@Deprecated(
        message = "Will be removed in later releases.",
        replaceWith = ReplaceWith("CustomerResource", imports = arrayOf("org.ostelco.prime.client.api.resources.CustomerResource")),
        level = WARNING)
@Path("/subscription")
class SubscriptionResource(private val dao: SubscriberDAO) {

    @GET
    @Path("activePseudonyms")
    @Produces(MediaType.APPLICATION_JSON)
    fun getActivePseudonyms(@Auth token: AccessTokenPrincipal?): Response {
        if (token == null) {
            return Response.status(Response.Status.UNAUTHORIZED)
                    .build()
        }

        return dao.getActivePseudonymForSubscriber(
                identity = Identity(id = token.name, type = "EMAIL", provider = token.provider))
                .fold(
                        { apiError -> Response.status(apiError.status).entity(asJson(apiError)) },
                        { pseudonym -> Response.status(Response.Status.OK).entity(pseudonym) })
                .build()
    }
}

@Path("/subscriptions")
class SubscriptionsResource(private val dao: SubscriberDAO) {

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    fun getSubscription(@Auth token: AccessTokenPrincipal?): Response {
        if (token == null) {
            return Response.status(Response.Status.UNAUTHORIZED)
                    .build()
        }

        return dao.getSubscriptions(
                identity = Identity(id = token.name, type = "EMAIL", provider = token.provider))
                .fold(
                        { apiError -> Response.status(apiError.status).entity(asJson(apiError)) },
                        { Response.status(Response.Status.OK).entity(asJson(it)) })
                .build()
    }
}
